kong/resty/dns_client/init.lua

```diff
@@ -0,0 +1,798 @@
+        ttl = ttl,
+      },
+    }
+    -- insert via the `:get` callback to prevent inter-process communication
+    cache:get(key, nil, function()
+      return answers, nil, ttl
+    end)
+  end
+
+  for name, address in pairs(hosts) do
+    name = string_lower(name)
+
+    if address.ipv4 then
+      insert_answer(name, TYPE_A, address.ipv4)
+      insert_last_type(cache, name, TYPE_A)
+    end
+
+    if address.ipv6 then
+      insert_answer(name, TYPE_AAAA, address.ipv6)
+      if not address.ipv4 or preferred_ip_type == TYPE_AAAA then
+        insert_last_type(cache, name, TYPE_AAAA)
+      end
+    end
+  end
+
+  return hosts
+end
+
+
+-- distinguish the worker_events sources registered by different new() instances
+local ipc_counter = 0
+
+function _M.new(opts)
+  if not opts then
+    return nil, "no options table specified"
+  end
+
+  -- parse resolv.conf
+  local resolv, err = utils.parse_resolv_conf(opts.resolv_conf, opts.enable_ipv6)
+  if not resolv then
+    log(WARN, "Invalid resolv.conf: ", err)
+    resolv = { options = {} }
+  end
+
+  -- init the resolver options for lua-resty-dns
+  local nameservers = (opts.nameservers and #opts.nameservers > 0) and
+                      opts.nameservers or resolv.nameservers
+  if not nameservers or #nameservers == 0 then
+    log(WARN, "Invalid configuration, no nameservers specified")
+  end
+
+  local r_opts = {
+    retrans     = opts.retrans or resolv.options.attempts or 5,
+    timeout     = opts.timeout or resolv.options.timeout or 2000, -- ms
+    no_random   = opts.no_random or not resolv.options.rotate,
+    nameservers = nameservers,
+  }
+
+  -- init the mlcache
+
+  -- maximum timeout for the underlying r:query() operation to complete
+  -- socket timeout * retrans * 2 calls for send and receive + 1s extra delay
+  local lock_timeout = r_opts.timeout / 1000 * r_opts.retrans * 2 + 1 -- s
+
+  local resty_lock_opts = {
+    timeout = lock_timeout,
+    exptimeout = lock_timeout + 1,
+  }
+
+  ipc_counter = ipc_counter + 1
+  local ipc_source = "dns_client_mlcache#" .. ipc_counter
+  local ipc = {
+    register_listeners = function(events)
+      -- The DNS client library will be required in globalpatches before Kong
+      -- initializes worker_events.
+      if not kong or not kong.worker_events then
+        return
+      end
+
+      local cwid = ngx.worker.id()
+      for _, ev in pairs(events) do
+        local handler = function(data, event, source, wid)
+          if cwid ~= wid then
+            ev.handler(data)
+          end
+        end
+
+        kong.worker_events.register(handler, ipc_source, ev.channel)
+      end
+    end,
+
+    broadcast = function(channel, data)
+      if not kong or not kong.worker_events then
+        return
+      end
+
+      local ok, err = kong.worker_events.post(ipc_source, channel, data)
+      if not ok then
+        log(ERR, "failed to post event '", ipc_source, "', '", channel, "': ", err)
+      end
+    end,
+  }
```

Comment:
Here `ipc_source` value is a upvalue, so we could not convert it to a module constant.

Why we need to use `ipc_source`, its that dns_client supports reinitialization mechanism. For different caller, we also need to reinit mlcache, which needs you to create a new mlcache object via  `mlcache:new()`. If you init mlcache with same `ipc` register functions, it will add this mlcache's event into same source of `kong.worker_events`. Then if someone triggers mlcache's event, the event will be broadcasted to all registered event source.

I tried to unregister the event of previous dns client object from kong.worker_events if we reinitialize `dns client->mlcache`, but it does not success. This process requires a significant amount of effort, so I believe we should mark it as `TODO` here just using current `ipc_source` to distinguish different sources.

Reply:
Consider our development schedule I think this solution is acceptable.